docs/proposals/20210716-async-azure-resource-creation-deletion.md

```diff
@@ -0,0 +1,365 @@
+type AzureMachineStatus struct {
+    /*
+      Other fields omitted for brevity    
+    */
+
+    // LongRunningOperationStates saves the states for Azure long-running operations so they can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationStates Futures `json:"longRunningOperationState,omitempty"`
+}
+
+// AzureMachinePool defines the observed state of AzureMachinePool.
+type AzureMachinePool struct {
+    /*
+      Other fields omitted for brevity    
+    */
+
+    // LongRunningOperationStates saves the states for Azure long-running operations so they can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationStates Futures `json:"longRunningOperationState,omitempty"`
+}
+
+// AzureMachinePoolMachineStatus defines the observed state of AzureMachinePoolMachine.
+type AzureMachinePoolMachine struct {
+    /*
+      Other fields omitted for brevity
+    */
+
+    // LongRunningOperationStates saves the states for Azure long-running operations so they can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationStates Futures `json:"longRunningOperationState,omitempty"`
+}
+
+```
+
+#### <a name='ProposedControllerChanges'></a>Proposed Controller Changes
+
+* Add an `AzureClientTimeout` which is the duration after which an Azure operation is considered a long running operation which should be handled asynchronously. Proposed starting value is 5 seconds.
+* For each Azure API call which returns a Future, wait for the operation to be completed for the above timeout duration. If the operation is not completed within the timeout duration, set Future of that resource in `LongRunningOperationStates` with the marshalled future data.
+* Set conditions at the end of each controller loop that describe the current state of the object and its associated Azure resources.
+
+For each Azure service, this is what the new asynchronous reconcile and delete flows will look like:
+
+##### Reconcile
+
+![Figure 1](./images/async-reconcile.png)
+
+##### Delete
+
+![Figure 2](./images/async-delete.png)
+
+#### <a name='ProposedNewConditions'></a>Proposed New Conditions
+
+The existing conditions before this proposal can be seen [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/v0.5.0/api/v1alpha4/conditions_consts.go). Note that these existing conditions will be left unchanged, and are purposeffully left out below.
+
+Part of the proposed changes is to add new conditions for Azure CRDs. More granular conditions, paired with more responsive controllers, will allow for better visibility into the state of each resource. Initially, the following conditions will be added:
+
+```go
+// Azure Services Conditions and Reasons.
+const (
+    // ResourceGroupReadyCondition means the resource group exists and is ready to be used.
+    ResourceGroupReadyCondition clusterv1.ConditionType = "ResourceGroupReady"
```

Comment:
I'm not sure if "Ready" is the right word. Other options I considered are "Created", "Succeeded", and "Running". I thought ready was a good choice because it also applies for BYO resource cases where the resource doesn't actually get created by the controller. "Running" also seems awkward since some of these resources don't actually "run". Thoughts?

Reply:
I like "Ready". It's convenient to remember as it's a common to many existing resources. It might not be 100% correct, but I think it conveys the right idea.

Reply:
That's a lot of conditions. I wonder if there is some way we can remove toil in describing / setting these conditions. Do you think we could name them through convention rather than declaring all of them, or do you think that would take away from discoverability?

Reply:
I agree it takes a lot of space in the proposal... but in the actual code it will be in its own constants file and I do think it's worth listing them explicitly. In the proposal I can simplify by just defining the pattern and providing an example or two, would that be better?

Reply:
Nope. I just wanted us to discuss. I'm 100% good with listing them in the proposal and including the consts in the code.

Reply:
I actually ended up simplifying them by reusing the same Reasons across conditions.